### ISSUE TITLE
scons launcher and scons dist don't rebuild

### ISSUE THREAD

**Issue by mltony** [nvaccess/nvda#13372]
### TL;DR
`scons launcher` or `scons dist` commands don't  seem to be able to pick up code changes after first execution. In other words, they  can build  successfully for the first time, but they  cannot rebuild as they don't appear to see that code has changed.
As to why this can be useful, in my example, I have NVDA build environment on my personal laptop, but I am trying to debug an issue with MS WOrd, which is only installed on my work laptop. I cannot build NVDA on work laptop, nor can I install MS Word on my personal laptop. As a result I need to frequently repeat the cycle of making code change, rebuild and test on work laptop.
### Steps to reproduce:
The following steps use `scons launcher` command but all the same applies to `scons dist` as well.
1. Get a fresh copy of NVDA:
```
> git clone --recursive https://github.com/nvaccess/nvda.git
> ...
> cd nvda
```
2. Build launcher:
```
> scons launcher
... many lines of actual building process omitted
scons: done building targets.
```
3. Check that launcher was built successfully:
```
> output\nvda_snapshot_source-master-6310b86.exe
```
4. Adding the following lines in core.py in def main right before `app.MainLoop()`:
```
import tones
tones.beep(1000, 1000)
```
5. Running NVDA from source to make sure beep can be heard:
```
> runnvda.bat
```
6. Rebuild launcher:
```
> scons launcher
...
scons: `launcher' is up to date.
scons: done building targets.
```
7. It appears scons didn't see any changes and didn't do anything, but let's try to run launcher again to see whether the beep can be heard:
```
$ output\nvda_snapshot_source-master-6310b86.exe
```

### Actual behavior:
`scons launcher` doesn't see code changes and does not rebuild after code change. After my attempt to rebuild, no beep is heard when launching launcher.
### Expected behavior:
`scons launcher` should be able to see code changes and rebuild launcher with code changes. Beep should be heard when launching launcher.

#### NVDA installed/portable/running from source:
Running from source
#### NVDA version:
Commit 6310b8661c7bc2f3d551fbe157a9e83a179e8b10
#### Windows version:
Windows 11
#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/A

**Comment by mltony** [comment]
In the end I found a workaround to this issue (thanks @josephsl for suggesting this in #10833):
1. Commit changes to a local commit.
2. Run:
```
scons --clean
git reset --hard
scons launcher
```
Still it would be great if scons could work without manually cleaning the whole repository.
cc: @lukaszgo1

**Comment by lukaszgo1** [comment]
In fact it is unnecessary to clean the whole repository - you just need to delete the `dist` and `output` directories.

**Comment by mltony** [comment]
Nope, scons launcher fails after that.

**Comment by mltony** [comment]
@lukaszgo1, Actually you're right, it appears to work. I was sure I tried that before though but must be in combination with scons --clean that apparently broke the build. So that is a valid work around. I would keep this issue open though, since I would expect a build system to track file changes and not require manual cleanup.

**Comment by hwf1324** [comment]
I spent a whole day investigating this issue. I'm sorry, but I actually have very little knowledge about building systems, so I could only look through some official documentation to find possible solutions.

Below are some of my survey results:

In SCons, using the `--tree` parameter allows you to observe the reference to the dependency `core.py`.

But in the result of the `--debug=prepare` parameter, it is missing again.

I don't understand why SCons doesn't check for changes in these files, the scripts in this repo are too complex for me.

Finally, I submitted a PR #15877, using the `AlwaysBuild` function to force the `dist` to be rebuilt every time the build is executed to solve this issue.

But there are still some flaws, I don't know how to make py2exe only replace the modified parts during construction
@LeonarddeR What is your opinion?
